remote_execution/oss/re_grpc/src/client.rs

```diff
@@ -471,72 +582,183 @@ impl REClient {
         metadata: RemoteExecutionMetadata,
         request: UploadRequest,
     ) -> anyhow::Result<UploadResponse> {
-        let mut client = self.grpc_clients.cas_client.clone();
-
-        let files_with_digest: Vec<Request> = request
-            .files_with_digest
-            .unwrap_or_default()
-            .into_try_map(|x| {
-                anyhow::Ok(Request {
-                    digest: Some(tdigest_to(x.digest)),
-                    // FIXME: This could do a lot of blocking reads
-                    data: fs_util::read(&x.name)?,
-                    compressor: compressor::Value::Identity as i32,
-                })
-            })?;
+        // NOTE if we stop recording blob_hashes, we can drop out a lot of allocations.
+        let mut upload_futures: Vec<BoxFuture<anyhow::Result<Vec<String>>>> = vec![];
+        let max_msg_size = self.capabilities.max_msg_size;
+        let client_uuid = uuid::Uuid::new_v4().to_string();
```

Comment:
We don't really validate (or expect) that we don't have duplicate blobs in the input here, I think you should generate a new UUID for each upload (i.e. in the place where you generate `resource_name`). This way, we don't have to worry about potentially using the same UUID for two separate uploads of the same blob, and it's not really all that expensive anyway.